### PR TITLE
[FLINK-30688][followup] Disable Kryo fallback for tests in flink-ml-lib

### PR DESCRIPTION
**Pull request by jiangxin369**
## What is the purpose of the change

Disable Kryo fallback in Flink ML tests to ensure that most Flink ML algorithms are not using Kryo.

With this PR, only the following algorithms are still using Kryo due to 2 types of reasons.

Iteration serialization issues:

- LogisticRegression
- OnlineLogisticRegression
- KMeans
- OnlineKMeans
- LinearSVC
- LinearRegression

Input rows with non-determined type:

- HashingTF

## Brief change log

*(for example:)*
  - Configures StreamExecutionEnvironment using env.getConfig().disableGenericTypes() in all Flink ML tests.
  - Add TypeInfoFactory for QuantileSummary
  - Add TypeInfoFactory for Vector

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-ml-lib/src/main/java/org/apache/flink/ml/common/util/quantile/QuantileSummary.java [hunk 1]
```diff
@@ -395,12 +400,14 @@ public QueryResult(int index, long minRankAtIndex, double percentile) {
...
      */
-    private static class StatsTuple implements Serializable {
+    public static class StatsTuple implements Serializable {
         private static final long serialVersionUID = 1L;
-        private final double value;
+        private double value;
         private long g;
         private long delta;
 
+        public StatsTuple() {}
```
Comment (lindong28) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/common/util/quantile/QuantileSummary.java:409]: Is this change needed to make this class POJO? If yes, then why its member variable is private?

File: flink-ml-lib/src/main/java/org/apache/flink/ml/common/util/quantile/QuantileSummary.java [hunk 2]
```diff
@@ -36,6 +37,7 @@
  * algorithm proposed in the paper: "Space-efficient Online Computation of Quantile Summaries" by
  * Greenwald, Michael and Khanna, Sanjeev. (https://doi.org/10.1145/375663.375670)
  */
+@TypeInfo(QuantileSummaryTypeInfoFactory.class)
```
Comment (lindong28) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/common/util/quantile/QuantileSummary.java:40]: Would it be simpler to put both `QuantileSummary` and `QuantileSummaryTypeInfoFactory` directly under the package 
`common/util`?

Note that `QuantileSummaryTypeInfoFactory` should be an internal class that users should not need to access directly.

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/lsh/LSHModel.java [hunk 3]
```diff
@@ -373,34 +395,45 @@ public TopKFunction(String distCol, int numNearestNeighbors) {
...
-                }
+        public List<Row> add(Row value, List<Row> accumulator) {
+            if (topKRows == null) {
+                topKRows = new PriorityQueue<>(numNearestNeighbors, new DistColComparator(distCol));
+                topKRows.addAll(accumulator);
             }
-            accumulator.add(value);
-            return accumulator;
+            insert(value, topKRows);
+            return new ArrayList<>(topKRows);
```
Comment (lindong28) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/feature/lsh/LSHModel.java:409]: It would be pretty inefficient to copy the entire list every time we add a value.

How about we use `PriorityQueue<Row>` as the accumulator and let `getResult(List<Row> accumulator)` return `List<ROW>`? I think we will only need to serialize the output of `getResult` but not the accumulator itself.

**Review by lindong28 — COMMENTED** [review]
Thanks for the PR. Left some comments below.

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/lsh/LSHModel.java (hunk 3 above)
Comment (lindong28) [reply flink-ml-lib/src/main/java/org/apache/flink/ml/feature/lsh/LSHModel.java:409]: Note that if the aggregator needs to be serializable by either `kryo` or our custom serializer, we can define a specific container class that has the `PriorityQueue<ROW>` as member variable and define a custom serializer for it.

File: flink-ml-lib/src/main/java/org/apache/flink/ml/common/util/quantile/QuantileSummary.java (hunk 1 above)
Comment (jiangxin369) [reply flink-ml-lib/src/main/java/org/apache/flink/ml/common/util/quantile/QuantileSummary.java:409]: No, it isn't. We don't need to make its member variable public because I define the `TypeInfomation` for `StatsTuple` with `Types.POJO`. We need to

File: flink-ml-lib/src/main/java/org/apache/flink/ml/common/util/QuantileSummary.java [hunk 4]
```diff
@@ -367,6 +373,26 @@ public double getRelativeError() {
         return relativeError;
     }
 
+    public int getCompressThreshold() {
```
Comment (lindong28) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/common/util/QuantileSummary.java:376]: Is this needed to make this class POJO? Would it be simpler to still keep these member variables `public`?

File: flink-ml-lib/src/main/java/org/apache/flink/ml/common/typeinfo/QuantileSummaryTypeInfoFactory.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        fields.put("relativeError", Types.DOUBLE);
+        fields.put("compressThreshold", Types.INT);
+        fields.put("count", Types.LONG);
+        fields.put("sampled", Types.LIST(TypeInformation.of(QuantileSummary.StatsTuple.class)));
+        fields.put("headBuffer", Types.LIST(Types.DOUBLE));
+        fields.put("compressed", Types.BOOLEAN);
+    }
+
+    private static final TypeInformation<QuantileSummary> TYPE_INFO =
+            Types.POJO(QuantileSummary.class, fields);
```
Comment (lindong28) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/common/typeinfo/QuantileSummaryTypeInfoFactory.java:50]: Can we just use `TypeInformation.of(QuantileSummary.class)` here after making `QuantileSummary` POJO, instead of having to explicitly specifying the `fields` above?

And do we still need `QuantileSummaryTypeInfoFactory` if `QuantileSummary` is POJO?

**Review by lindong28 — COMMENTED** [review]
Thanks for the update. Left some comments below.

File: flink-ml-lib/src/main/java/org/apache/flink/ml/common/typeinfo/QuantileSummaryTypeInfoFactory.java (hunk 5 above)
Comment (jiangxin369) [reply flink-ml-lib/src/main/java/org/apache/flink/ml/common/typeinfo/QuantileSummaryTypeInfoFactory.java:50]: According to [1], a POJO must not contain fields that are not supported by registered serializers, while `QuantileSummary` has fields like `List<Double> headBuffer`, and `List` has no registered serializer and is treated as a generic type in Flink. So we still have to register the type info with `Types.POJO` and explicitly specify the fields.

[1] https://nightlies.apache.org/flink/flink-docs-master/docs/dev/datastream/fault-tolerance/serialization/types_serialization/

File: flink-ml-lib/src/main/java/org/apache/flink/ml/common/util/QuantileSummary.java (hunk 4 above)
Comment (jiangxin369) [reply flink-ml-lib/src/main/java/org/apache/flink/ml/common/util/QuantileSummary.java:376]: No, it is for unittests currently. But I didn't annotate it with `VisibleForTesting` because I think it is reasonable for users to read the information of a `QuantileSummary`.

**Comment by lindong28** [comment]
Thanks for the update. LGTM.
